arrow-fx-coroutines/README.MD

```diff
@@ -1,561 +1,561 @@
-    throw RuntimeException("Boom!")
-  }
+  // ...
 }
 ```
 
-Additionally a `Schedule` can also produce an output, let's compose our simple `recurs` Schedule to `recur` every `10.seconds` in a `spaced` fashion and collect all outputs along the way.
+A `CancelToken` is a `inline class` wrapper around `suspend () -> Unit` which is used to model cancellation token. These are used to define cancellable operations in a typed manner.
 
-```kotlin:ank
-fun <A> schedule(): Schedule<A, List<A>> = Schedule {
-  (recurs<A>(10) and spaced(10.seconds)) zipRight collect()
-}
+It has two implementations, `Uncancellable` and `DefaultConnection`
+
+#### Uncancellable
 
-suspend fun main(): Unit {
-  val count = Atomic(0)
+The uncancellable implementation has a no-op `cancel` method, and always returns `false` upon checking `isCancelled`.
+Which subsequently means that all cancellable operations become uncancellable when run on this `CoroutineContext`.
+Some optimisations could be implemented when an `uncancellable` context is detected.
 
-  val history: List<Int> = repeat(schedule<Int>()) {
-    println("Incrementing the ref")
-    count.update(Int::inc)
+#### DefaultConnection
+
+The `DefaultConnection` keeps track of all `CancelToken` registered for a given `SuspendConnection`.
+It does so by keeping all `CancelToken` in a FIFO stack, and running them in order when cancelled.
+Since `SuspendConnection` is a FIFO stack, we can register a token before running a cancellable operation,
+and pop the token after the cancellable operation has finished so
+we don't run a `CancelToken` when it's not necessary anymore since that could lead to (undefined) weird behavior. This is useful for example in `parMap2` where we want to launch two parallel operations, and push their cancellation system onto to the original `SuspendConnection` and `pop` it when both finish.
+
+### Cancellation
+
+So what does it actually mean to cancel? Currently in Arrow 0.11.0, cancellation means that the code stops running (or in other words suspends forever and `never` returns).
+This is implemented as a `cancelBoundary`, and with all the understanding of the low-level Kotlin machinery we have now can be implemented as follows.
+
+```kotlin
+suspend fun cancelBoundary(): Unit =
+  suspendCoroutineUninterceptedOrReturn { cont ->
+    if (cont.context[SuspendConnection]?.isCancelled() == true) COROUTINE_SUSPENDED
+    else Unit
   }
-}
 ```
 
-## Parallel Ops
+So a `cancelBoundary` checks if the `SuspendConnection#isCancelled` and if so it suspends forever and never invokes the callback, effectively this means it stops running the code.
+In the case the `SuspendConnection` isnt' cancelled it immediately returns `Unit`. Here we use `suspendCoroutineUninterceptedOrReturn` since we can return immediately when no cancellation signal was triggered, and we don't want to intercept or schedule but we have to check the `Continuation#context` to be able to access the `SuspendConnection`.
```

Comment:
```suggestion
In the case the `SuspendConnection` isn't cancelled, it immediately returns `Unit`. Here we use `suspendCoroutineUninterceptedOrReturn` since we can return immediately when no cancellation signal was triggered, and we don't want to intercept or schedule but we have to check the `Continuation#context` to be able to access the `SuspendConnection`.
```